keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,429 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node labels
+to strictly those that are standardized through KEP-1659.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+standard topology labels are available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
+* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.
+* `PodTopologyLabels` admission will overwrite `topology.k8s.io/*` labels on Pods.
+* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.
+The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.
+* The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.
+At this point we will overwrite Pod labels in Binding that are allowed to be exposed via Downward API.
+* For exposing additional node labels, at the discretion of the cluster admin, a mutating admission webhook can be used to mutate labels of the `pods/binding` subresource.
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests will be added for:
+* New `PodTopologyLabels` admission plugin
+* Binding REST implementation
+
+Unit tests will also ensure behavior is exercised when the feature gate is enabled.
+
+##### Integration tests
+
+Integration tests will be added to test the following behavior:
+* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
+* Topology labels can be expressed in Pod Downward API
+* Node labels outside standard topology labels are disallowed
+* Topology labels are empty if underlying Node does not specify a topology
+
+Integration tests will also ensure behavior is exercised when the feature gate is enabled.
+
+##### e2e tests
+
+E2E tests will be added to test the following scenarios:
+* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
+* A Pod using downward API to retrieve the underlying topology information about the Node
+* A Pod attempting to use downward API to retrieve Node labels that are not the standard topology labels
```

Comment:
Might be worth adding a test case of consuming topology labels via downward API when the some labels are not set